src/librustc_mir/monomorphize/deduplicate_instances.rs

```diff
@@ -0,0 +1,174 @@
+// Copyright 2014 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use rustc_data_structures::indexed_vec::IndexVec;
+use rustc::ty::{self, TyCtxt, Ty, TypeFoldable, Instance, ParamTy};
+use rustc::ty::fold::TypeFolder;
+use rustc::ty::subst::Kind;
+use rustc::middle::const_val::ConstVal;
+use rustc::mir::{Mir, Rvalue, Promoted, Location};
+use rustc::mir::visit::{Visitor, TyContext};
+
+/// Replace substs which arent used by the function with TyError,
+/// so that it doesnt end up in the binary multiple times
+pub(crate) fn collapse_interchangable_instances<'a, 'tcx>(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    mut inst: Instance<'tcx>
+) -> Instance<'tcx> {
+    info!("replace_unused_substs_with_ty_error({:?})", inst);
+
+    if inst.substs.is_noop() || !tcx.is_mir_available(inst.def_id()) {
+        return inst;
+    }
+    match inst.ty(tcx).sty {
+        ty::TyFnDef(def_id, _) => {
+            //let attrs = tcx.item_attrs(def_id);
+            if tcx.lang_items().items().iter().find(|l|**l == Some(def_id)).is_some() {
+                return inst; // Lang items dont work otherwise
+            }
+        }
+        _ => return inst, // Closures dont work otherwise
+    }
+
+    let used_substs = used_substs_for_instance(tcx, inst);
+    inst.substs = tcx._intern_substs(&inst.substs.into_iter().enumerate().map(|(i, subst)| {
+        if let Some(ty) = subst.as_type() {
+            let ty = if used_substs.substs.iter().find(|p|p.idx == i as u32).is_some() {
+                ty.into()
+            } else if let ty::TyParam(ref _param) = ty.sty {
+                //^ Dont replace <closure_kind> and other internal params
+                if false /*param.name.as_str().starts_with("<")*/ {
+                    ty.into()
+                } else {
+                    tcx.sess.warn(&format!("Unused subst for {:?}", inst));
+                    tcx.mk_ty(ty::TyNever)
+                }
+            } else {
+                // Can't use TyError as it gives some ICE in rustc_trans::callee::get_fn
+                tcx.sess.warn(&format!("Unused subst for {:?}", inst));
+                tcx.mk_ty(ty::TyNever)
+            };
+            Kind::from(ty)
+        } else {
+            (*subst).clone()
+        }
+    }).collect::<Vec<_>>());
+    info!("replace_unused_substs_with_ty_error(_) -> {:?}", inst);
+    inst
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct UsedSubsts {
```

Comment:
I don't think we should call this `UsedSubsts`, but rather `UsedParameters` -- I think of *substs* as being the set of values that we use to substitute in. e.g., if you have `struct Vec<T>`, the `T` is not a subst but rather a generic parameter. In contrast, the `u32` in `Vec<u32>` is a subst.